### PR TITLE
Update nodejs versions

### PR DESCRIPTION
**Pull request by luiscoms**

### CODE REVIEW

File: centos7-s2i-nodejs.json [hunk 1]
```diff
@@ -166,76 +152,135 @@
...
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-7.5.0,nodejs-latest",
+              "supports":"nodejs:7,nodejs:7.5,nodejs",
+              "version": "7.5.0",
+              "sampleRepo": "https://github.com/ryanj/pillar-base.git"
+            },
+            "from": {
+              "kind": "DockerImage",
+              "name": "ryanj/centos7-s2i-nodejs:6.9.5"
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: should this be 7.5.0?

File: centos7-s2i-nodejs.json [hunk 2]
```diff
@@ -166,76 +152,135 @@
...
+            }
+          },
+          {
+            "name": "latest",
+            "annotations": {
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-6.9.5",
+              "supports":"nodejs:6,nodejs:6.9,nodejs",
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: 7?

File: centos7-s2i-nodejs.json [hunk 3]
```diff
@@ -166,76 +152,135 @@
...
+              "kind": "DockerImage",
+              "name": "ryanj/centos7-s2i-nodejs:6.9.5"
+            }
+          },
+          {
+            "name": "latest",
+            "annotations": {
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-6.9.5",
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: 7?

File: centos7-s2i-nodejs.json [hunk 4]
```diff
@@ -166,76 +152,135 @@
...
+              "supports":"nodejs:6,nodejs:6.9,nodejs",
+              "sampleRepo": "https://github.com/ryanj/pillar-base.git"
+            },
+            "from": {
+              "kind": "ImageStreamTag",
+              "name": "7.5.0"
+            }
+          },
+          {
+            "name": "current",
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: why do we need current *and* latest?

File: centos7-s2i-nodejs.json [hunk 5]
```diff
@@ -166,76 +152,135 @@
...
+            "annotations": {
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-6.9.5,nodejs-current",
+              "supports":"nodejs:6,nodejs:6.9,nodejs",
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: 7

File: centos7-s2i-nodejs.json [hunk 6]
```diff
@@ -166,76 +152,135 @@
...
+              "kind": "ImageStreamTag",
+              "name": "7.5.0"
+            }
+          },
+          {
+            "name": "current",
+            "annotations": {
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-6.9.5,nodejs-current",
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: 7

File: centos7-s2i-nodejs.json [hunk 7]
```diff
@@ -166,76 +152,135 @@
...
           {
-            "name": "6.3.1",
+            "name": "7.5",
             "annotations": {
               "description": "Build and run NodeJS applications",
               "iconClass": "icon-nodejs",
-              "tags": "builder,nodejs,nodejs-6.3.1,nodejs-latest",
-              "supports":"nodejs:6,nodejs:6.2,nodejs",
-              "version": "6.3.1",
+              "tags": "builder,nodejs,nodejs-6.9.5,nodejs-latest",
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: 7?

File: centos7-s2i-nodejs.json [hunk 8]
```diff
@@ -166,76 +152,135 @@
...
           {
-            "name": "current",
+            "name": "7",
             "annotations": {
               "description": "Build and run NodeJS applications",
               "iconClass": "icon-nodejs",
-              "tags": "builder,nodejs,nodejs-6.3.1,nodejs-current",
-              "supports":"nodejs:6,nodejs:6.2,nodejs",
+              "tags": "builder,nodejs,nodejs-7.5.0",
+              "supports":"nodejs:7,nodejs:6.5,nodejs",
```
Comment (bparees) [inline centos7-s2i-nodejs.json]: 7.5?

File: image-streams-candidate.json [hunk 9]
```diff
@@ -168,74 +154,133 @@
...
           {
-            "name": "current",
+            "name": "7",
             "annotations": {
               "description": "Build and run NodeJS applications",
               "iconClass": "icon-nodejs",
-              "tags": "builder,nodejs,nodejs-6.3.1,nodejs-current",
-              "supports":"nodejs:6,nodejs:6.2,nodejs",
+              "tags": "builder,nodejs,nodejs-7.5.0",
+              "supports":"nodejs:7,nodejs:6.5,nodejs",
```
Comment (bparees) [inline image-streams-candidate.json]: 7.5?

File: image-streams-candidate.json [hunk 10]
```diff
@@ -168,74 +154,133 @@
...
           {
-            "name": "6.3.1",
+            "name": "7.5",
             "annotations": {
               "description": "Build and run NodeJS applications",
               "iconClass": "icon-nodejs",
-              "tags": "builder,nodejs,nodejs-6.3.1,nodejs-latest",
-              "supports":"nodejs:6,nodejs:6.2,nodejs",
-              "version": "6.3.1",
+              "tags": "builder,nodejs,nodejs-6.9.5,nodejs-latest",
```
Comment (bparees) [inline image-streams-candidate.json]: 7?

File: image-streams-candidate.json [hunk 11]
```diff
@@ -168,74 +154,133 @@
...
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-7.5.0,nodejs-latest",
+              "supports":"nodejs:7,nodejs:7.5,nodejs",
+              "version": "7.5.0",
+              "sampleRepo": "https://github.com/ryanj/pillar-base.git"
+            },
+            "from": {
+              "kind": "DockerImage",
+              "name": "ryanj/centos7-s2i-nodejs:6.9.5"
```
Comment (bparees) [inline image-streams-candidate.json]: 7.5.0?

File: image-streams-candidate.json [hunk 12]
```diff
@@ -168,74 +154,133 @@
...
+            }
+          },
+          {
+            "name": "latest",
+            "annotations": {
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-6.9.5",
```
Comment (bparees) [inline image-streams-candidate.json]: 7

File: image-streams-candidate.json [hunk 13]
```diff
@@ -168,74 +154,133 @@
...
+              "supports":"nodejs:6,nodejs:6.9,nodejs",
```
Comment (bparees) [inline image-streams-candidate.json]: 7

File: image-streams-candidate.json [hunk 14]
```diff
@@ -168,74 +154,133 @@
...
+              "kind": "ImageStreamTag",
+              "name": "7.5.0"
+            }
+          },
+          {
+            "name": "current",
+            "annotations": {
+              "description": "Build and run NodeJS applications",
+              "iconClass": "icon-nodejs",
+              "tags": "builder,nodejs,nodejs-6.9.5,nodejs-current",
```
Comment (bparees) [inline image-streams-candidate.json]: 7

File: image-streams-candidate.json [hunk 15]
```diff
@@ -168,74 +154,133 @@
...
+              "supports":"nodejs:6,nodejs:6.9,nodejs",
```
Comment (bparees) [inline image-streams-candidate.json]: 7

File: image-streams-candidate.json [hunk 16]
```diff
@@ -168,74 +154,133 @@
...
+              "sampleRepo": "https://github.com/ryanj/pillar-base.git"
+            },
+            "from": {
+              "kind": "ImageStreamTag",
+              "name": "7.5.0"
+            }
+          },
+          {
+            "name": "current",
```
Comment (bparees) [inline image-streams-candidate.json:260]: why current+latest?

File: nodejs.org/Dockerfile.onbuild [hunk 17]
```diff
@@ -11,7 +11,7 @@ EXPOSE 8080
 # See https://docs.npmjs.com/misc/scripts, and your repo's package.json
 # file for possible values of NPM_RUN
 ENV NPM_RUN=start \
-    NODE_VERSION=6.3.1 \
+    NODE_VERSION= \
```
Comment (bparees) [inline nodejs.org/Dockerfile.onbuild:14]: 6.9.5?

**Review by bparees — CHANGES_REQUESTED** [review]
A bunch of the new imagestreams reference ryanj's dockerhub repo, but we don't have a way to push there, so those imagestreams aren't actually going to be able to import anything.  

Also for your "latest/current" imagestreams, I recommend being version agnostic so they don't have to be updated every time what they point to changes, see how we handle it here for example:
https://github.com/openshift/origin/blob/master/examples/image-streams/image-streams-centos7.json#L18-L30

vs

https://github.com/openshift/origin/blob/master/examples/image-streams/image-streams-centos7.json#L65-L78

**Comment by luiscoms** [comment]
@bparees Well, I updated these versions with the script provided on the project with `hack/rebuild.sh`, that update all versions on `Makefile`.

Then, I update the image streams adding the `7.5.0` to the latest and current tag and `6.9.5`.

This image exists on dockerhub?

**Comment by luiscoms** [comment]
@bparees Any questions?

**Comment by bparees** [comment]
@luiscoms it does answer the question about current vs latest in that you didn't introduce them (but i still don't think it makes sense to have both)

but based on way you added 7.5 it looks like the search/replace logic updated a bunch of stuff to 6.95 instead of 7.5, so it still needs to be cleaned up.  I imagine the script replaced all the 6.3 references w/ 6.95, so your new "7" imagestreams didn't then get updated properly since they no longer contained 6.3.

**Comment by luiscoms** [comment]
@bparees I will replace them but my question is if I just replace the version or the namespace too.
These image exists on dockerhub?

**Comment by bparees** [comment]
> @bparees I will replace them but my question is if I just replace the version or the namespace too.
These image exists on dockerhub?

they don't exist on dockerhub and we don't have any particular plans to build/push them to dockerhub.  if you're interested, you can push them to your dockerhub org and update the imagestreams to point there, or we can see if @ryanj is still interested in building the new versions and pushing them to his org.

**Comment by ryanj** [comment]
@bparees I've been meaning to roll out a 7.x image. I'd be happy to fix the automation scripts to cover this. Do builds still need to be run using docker-1.09?

**Comment by bparees** [comment]
> Do builds still need to be run using docker-1.09?

no, you can push w/ newer docker versions now and openshift should work fine.

**Comment by luiscoms** [comment]
@bparees I just update the PR with your considerations and added support to build on Mac OS

File: hack/build.sh [hunk 18]
```diff
@@ -25,23 +25,21 @@ trap "rm -f ${DOCKERFILE}.${version}" SIGINT SIGQUIT EXIT
...
 # Install the docker squashing tool[1] and squash the result image
 # [1] https://github.com/goldmann/docker-squash
 function squash {
-  # FIXME: We have to use the exact versions here to avoid Docker client
-  #        compatibility issues
-  easy_install -q --user docker_py==1.6.0 docker-squash==1.0.0rc6
+  pip install -q --user -U docker-squash
   base=$(awk '/^FROM/{print $2}' $1)
-  ${HOME}/.local/bin/docker-squash -f $base ${IMAGE_NAME}:${version}
+  $(python -m site --user-base)/bin/docker-squash -f $base ${IMAGE_NAME}:${version}
```
Comment (bparees) [inline hack/build.sh]: you're welcome to just delete the squashing logic entirely, we're moving away from it.

File: nodejs.org/Dockerfile.onbuild (hunk 17 above)
Comment (bparees) [reply nodejs.org/Dockerfile.onbuild:14]: is there a reason this has no value?

Comment (luiscoms) [reply nodejs.org/Dockerfile.onbuild:14]: The value is filled on build time

**Comment by luiscoms** [comment]
@bparees I removed the squash logic

File: nodejs.org/Dockerfile.onbuild (hunk 17 above)
Comment (bparees) [reply nodejs.org/Dockerfile.onbuild:14]: why not remove it in the other Dockerfiles then?

Comment (bparees) [reply nodejs.org/Dockerfile.onbuild:14]: I'm just aiming for consistency here :)

Comment (luiscoms) [reply nodejs.org/Dockerfile.onbuild:14]: Well, I can remove there too

Comment (bparees) [reply nodejs.org/Dockerfile.onbuild:14]: that works.

**Comment by luiscoms** [comment]
@bparees Is it ok now?

**Comment by bparees** [comment]
@luiscoms i'm validating your branch locally

**Comment by luiscoms** [comment]
@ryanj @bparees  Thesse images will be published on Dockerhub or not?

**Pull request merged**